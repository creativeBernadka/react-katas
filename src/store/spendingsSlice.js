import { createSlice } from "@reduxjs/toolkit";

export const spendingsSlice = createSlice({
  name: "spendings",
  initialState: [],
  reducers: {
    addNewSpending: (state, action) => {
      const allIds = state.map(({ id }) => id);
      const newIndex = allIds.indexOf(action.payload.id);
      if (newIndex > -1) {
        state[newIndex] = action.payload;
      } else {
        state.push(action.payload);
      }
    },
  },
});

export const { addNewSpending } = spendingsSlice.actions;

export const getAllSpendings = (state) => state.spendings;

export const getCurrentNumberOfSpendings = (state) => state.spendings.length;

export const getAllDistinctDates = (state) => {
  const dates = state.spendings.map(({ date }) => date);
  return [...new Set(dates)];
};

export const getAllSpendingsByDate = (date) => (state) => {
  return state.spendings.filter(
    ({ date: spendingsDate }) => spendingsDate === date
  );
};

export const getSpendingById = (id) => (state) => {
  return state.spendings.filter(({ id: spendingId }) => id === spendingId)[0];
};

export default spendingsSlice.reducer;
