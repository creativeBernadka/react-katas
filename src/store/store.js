import { configureStore } from '@reduxjs/toolkit';
import spendingsReducer from './spendingsSlice';

export default configureStore({
  reducer: {
    spendings: spendingsReducer,
  },
});