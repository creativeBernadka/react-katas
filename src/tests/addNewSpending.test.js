import { spendingsSlice } from "../store/spendingsSlice";

describe("addNewSpending", () => {

  it("correctly adds new spending", () => {
    const initialState = {
      spendings: [
        {
          date: "2020-12-05",
          category: {
            name: "",
            subcategoryName: "",
          },
          spendingValues: [""],
          id: "",
        },
      ],
    };
    const {caseReducers: {addNewSpending}} = spendingsSlice
  
    const results = addNewSpending(initialState.spendings , {payload: {id: 1, date: "2021-01-20"}})
    expect(results).toHaveLength(2)

  });
});

