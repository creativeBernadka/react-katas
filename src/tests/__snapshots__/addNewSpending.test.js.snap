// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> renders correctly 1`] = `ShallowWrapper {}`;

exports[`addNewSpending renders correctly 1`] = `ShallowWrapper {}`;
